content/blog/2019-11-17-chlorophyll/index.md

```diff
@@ -0,0 +1,236 @@
+
+#### Modular Superoptimization
+A behavior of a program is specified by its sequence of instructions *P*, and its live region *L*. 
+Consider the resulting data stacks from executing *P*.
+
+<img src="stack_spec.png" width="500" >
+
+*P* changes the data stack from $\alpha$|$\beta$ to $\alpha$|$\gamma$. Assume $\alpha$|$\gamma$ is in the live region. We say *P'* is equivalent to *P* if *P'* produces $\alpha$|$\gamma$, and the stack pointers after executing *P* and *P'* point to the same location.
+
+Consider figure (b) above. Let *P'* be the sequence of instructions resulting in (b). Note that under this definition of equivalence, the resulting stacks in (a) and (b) are different. This is because the stack pointer is pointing to different places on each stack.
+
+In order to decrease the number of false negatives, the authors loosen the specification. Assume *P* results in a data stack of $\alpha$|$\gamma$. *P'* is equivalent to *P* if *P'* results in a data stack of $\delta$|$\alpha$|$\gamma$, where $\delta$ may be empty. Additionally, they place no restrictions on the stack pointer.
+
+This is valid because the stacks are circular, and therefore leaving garbage at the bottom of the stack will simply be overwritten as the stack pointer is incremented accordingly.
+
+This engenders additional optimization as well. Consider the corresponding stacks for a subtract operation.
+
+<img src="subexample.png" width="400" >
+
+Here, b and a are popped from the stack, subtracted, and the result is pushed back onto the stack. The *P* that results in only `b-a` left on the stack is 8 instructions long, with 3 instructions to remove the remaining garbage value `a` from the stack. However, as noted before, it is in fact legal to leave `a` at the bottom of the stack, thus saving 3 instructions. 
+
+#### Sliding Windows
+To scale superoptimization to real world programs, the authors employ the *sliding windows technique*.
+This algorithm merges superoptimizable units into a superoptimizable segment. It starts with an empty segment, superoptimizes the largest sequence possible from the input instruction sequence, and appends the result to the output, if valid.
+In the case where no valid segment is found from the superoptimizer, it appends the the first unit from the segment to the output, and attempts to build another sequence to superoptimize. 
+Then, it iterates until the instruction sequence is empty.
+
+The superoptimizer executes a binary search over programs given a cost model. It uses counterexample-guided inductive synthesis (CEGIS) to synthesize a program of cost k, and if one exists, to synthesize a program of cost k/2, and so on.
+
+When perusing papers that cite this one, we found the primary critique of the *sliding windows technique* to be the focus on local optimization and inability to make whole-program optimality guarantees. Other synthesis approaches such as Metasketches, claim to provide these guarantees.[[6]](#6)
```

Comment:
Citations before periods, as above. (Also, I don’t think you need to capitalize metasketches.)